### PR TITLE
Overhaul Windows install docs 

### PR DESCRIPTION
**Pull request by dra27**
- Reformatted sections of `INSTALL.adoc`
- Moved Cygwin instructions from `README.win32.adoc` to `INSTALL.adoc`
- Removed out-of-date warnings and instructions (Cygwin packages, etc.)
- Simplified the references to Windows ports by regarding 32/64-bit as a version, rather than a separate port (since Cygwin/MSVC/Mingw-w64 all work in both architectures)
- Documented the MSVC/Mingw-w64 bootstrapping of FlexDLL introduced in 4.03
- Added `tools/msvs-promote-path` to simplify the environment set-up for building MSVC
- Updated the instructions for MSVC to cover **all** (literally, all) supported versions of Visual Studio and the Windows SDK which can build OCaml

Sorry that this is so close to the release date - though no code is touched.

Alain hasn't released FlexDLL 0.35 yet - when he does, the git submodule should be moved forwards to that commit both in `trunk` and `4.03`.

### CODE REVIEW

**Comment by protz** [comment]
Not sure why... I'm happy to use Alain's binary. Is there a reason why I 
would want to bootstrap?

**Comment by dra27** [comment]
Potentially simpler build? It would also mean that the x64 distribution would have a 64-bit compiled version of flexlink, rather than a 32-bit one (see part of Alain's https://github.com/ocaml/ocaml/pull/388#issuecomment-168612078 for when this may be desirable). I did mean "may be" not "will be" though :smile:

File: README.win32.adoc [hunk 1]
```diff
@@ -36,373 +36,292 @@ Mingw-w64 have no such restrictions.
...
+   using a sources tarball, you can run:
++
+  git submodule update --init
+
+OCaml is then compiled as normal for the port you require, except that before
+compiling `world`, you must compile `flexdll`, i.e.:
+
+  make -f Makefile.nt flexdll world [bootstrap] opt opt.opt install
+
+ * `make -f Makefile.nt` will install FlexDLL by placing `flexlink.exe` (and the
```
Comment (alainfrisch) [inline README.win32.adoc]: Should this be `make -f Makefile.nt flexdll`?

Comment (dra27) [inline README.win32.adoc]: Oops - actually it should read `make -f Makefile.nt install` - it's describing the fact that a bootstrapped compilation also installs FlexDLL as part of OCaml.

**Comment by alainfrisch** [comment]
Should msvs-promote-path be installed?  (Yes, it might be weird to install a Cygwin specific script for a native install...)  Alternatively, one could direct people to download the tool from your repository directly.

**Comment by dra27** [comment]
Up to you - the only thing if it is installed would be whether it wants to be installed with that name (or prefixed `ocaml-`)? As far as I can tell, it's certainly not uncommon to use the native ports from Cygwin bash, so it's definitely not a silly thing to want to do!

**Comment by alainfrisch** [comment]
Do you plan to release msvs-tools as a standalone package?  If so, I think the best would be to refer to it as an (optional) dependency from the README (msvs-detect would also probably be useful), and let Windows binary packagers decide whether to bundle the tools with their installer.

(Another approach for the "link" issue would be to address it within flexlink itself.  This might be the most user friendly approach.)

**Comment by gasche** [comment]
Where are we on this? If we want to have it in 4.03, there should be a clear consensus to merge very soon (as in, the next day or so). I would personally be tempted to just blindly trust @dra27 on the documentation changes, but I can't assess the risk of the new instructions making it worse for someone than the old. Any opinion?

**Comment by dra27** [comment]
Sorry for the slow reply.
@alainfrisch - msvs-tools is as released as I was planning (which is to say it's on GitHub and has a tagged version history). I anticipate it being like [ocaml-autoconf](http://ocaml-autoconf.forge.ocamlcore.org), so something which you would tend to embed in other applications directly, rather than need installed as a dependency. The README kind of does refer to the tool already inasmuch as it recommends copying the script from `tools/` if you want to use it after building OCaml - would you think that's sufficient? In the bright OPAM-driven future, this becomes moot - OPAM will necessary have to deal with the environment in order to support switching between different architectures. I agree that moving all of this to flexlink would be best, though there are two caveats - we'd been discussing it back in either December or January before the code-freeze and concluded that was best left to 4.04 (or at least 4.03.1) since it requires a change to the OCaml Windows build system for the cases where ocamlc/ocamlopt call the linker directly. I'm very happy to help with this work post 4.03!
@gasche - although moderately awkward (for checksums and so on), presumably the sources tarball can be re-released under the same version number if the only thing altered is documentation?

**Comment by mshinwell** [comment]
I don't think we should re-release a version even if there is only a doc change.

**Comment by dra27** [comment]
There is a confidence-test in the new instructions with the msvs-promote-path command which would be handy to have run by people with a system set-up that is manually configured "the old way" (@damiendoligez, @alainfrisch) - running `eval $(tools/msvs-promote-path)` should simply display `link already refers to the Microsoft Linker` on stderr and change nothing, because the correct `link` is already in `PATH`. That would confirm that backwards compatibility of the instructions hasn't been broken.
There is of course one tiny code change which I made in `configure` to remove the dependency on the dos2unix Cygwin package - I have tested that on both Cygwin and Cygwin64, and I think it's fairly clear by inspection that the two lines are equivalent?

**Comment by gasche** [comment]
Indeed, the ball would be in @alainfrisch and @damiendoligez 's court.

Note that the plan for after the release is to switch to a shorter release cycle (as in: have another release in 2016), so missing this one wouldn't be as much of a problem.

**Comment by damiendoligez** [comment]
> running `eval $(tools/msvs-promote-path)` should simply display link already refers to the Microsoft Linker on stderr and change nothing

It works as advertised, except that `tools/msvs-promote-path` doesn't have execute permission. Can you fix this?

**Comment by dra27** [comment]
Done (also squashed the typo commit and rebased)

**Comment by dra27** [comment]
Congratulations on the release of 4.03.0! Just a small note, in case in the last-minute rush it's slipped through, that this needs merging on trunk.

**Comment by gasche** [comment]
Thanks for the reminder, I just merged in trunk.

**Pull request merged**